### PR TITLE
Regressor for RLDDM

### PR DESCRIPTION
**Pull request by madslupe**
Created a new class called HDDMrlRegressor that allows estimating instrumental learning with the RLDDM while also estimating the impact of continuous regressors. It builds on the HDDMRegressor, but adds a learning rate parameter and takes as input feedback on choices. A potential issue is that I also included alpha as a parameter for the normal HDDMRegressor, because I create alpha in base.py. Ideally this would be created within the HDDMrlRegressor class, but I didn't get that working when also letting regressors impact alpha.

### CODE REVIEW

File: hddm/likelihoods.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def wiener_like_contaminant(value, cont_x, v, sv, a, z, sz, t, st, t_min, t_max,
                                          logp=wiener_like_contaminant)
 
 def general_WienerCont(err=1e-4, n_st=2, n_sz=2, use_adaptive=1, simps_err=1e-3):
-    _like = lambda  value, cont_x, v, sv, a, z, sz, t, st, t_min, t_max, err=err, n_st=n_st, n_sz=n_sz, \
+    _like = lambda  value, cont_x, v, sv, a, z, sz, t, st,  t_min, t_max, err=err, n_st=n_st, n_sz=n_sz, \
```
Comment (twiecki) [inline hddm/likelihoods.py]: superfluous space.

File: hddm/models/hddm_regression.py [hunk 2]
```diff
@@ -22,7 +22,7 @@ def generate_wfpt_reg_stochastic_class(wiener_params=None, sampling_method='cdf'
                          'w_outlier': 0.1}
     wp = wiener_params
 
-    def wiener_multi_like(value, v, sv, a, z, sz, t, st, reg_outcomes, p_outlier=0):
+    def wiener_multi_like(value, v, sv, a, z, sz, t, st, alpha, reg_outcomes, p_outlier=0):
```
Comment (twiecki) [inline hddm/models/hddm_regression.py]: I don't think you need to add alpha to anything outside of your new class.

File: hddm/models/base.py [hunk 3]
```diff
@@ -743,6 +743,7 @@ def _create_wfpt_parents_dict(self, knodes):
         wfpt_parents['a'] = knodes['a_bottom']
         wfpt_parents['v'] = knodes['v_bottom']
         wfpt_parents['t'] = knodes['t_bottom']
+        wfpt_parents['alpha'] = knodes['alpha_bottom'] if 'alpha' in self.include else 0
```
Comment (twiecki) [inline hddm/models/base.py]: You can just overload __init__ in your class and add this there.

File: hddm/models/hddm_regression.py (hunk 2 above)
Comment (madslupe) [reply hddm/models/hddm_regression.py:25]: I've tried adding alpha directly in the HDDMrlRegressor class (see below for what I've tried), but then it does not allow running regressors onto alpha. 

from HDDMrlRegressor.py under class HDDMrlRegressor:

def _create_wfpt_knode(self, knodes):
        wfpt_parents = self._create_wfpt_parents_dict(knodes)
        wfpt_parents['alpha'] = knodes['alpha_bottom']
        return Knode(self.wfpt_reg_class, 'wfpt', observed=True,
                     col_name=['split_by', 'feedback', 'response', 'rt', 'q_init'],
                     reg_outcomes=self.reg_outcomes, **wfpt_parents)

    def _create_stochastic_knodes(self, include):
        # Create all stochastic knodes except for the ones that we want to replace
        # with regressors.
        knodes = super(HDDMrlRegressor, self)._create_stochastic_knodes(include.difference(self.reg_outcomes))
        knodes.update(self._create_family_normal(
                    'alpha', value=0, g_mu=0.2, g_tau=3**-2, std_lower=1e-10, std_upper=10, std_value=.1))

Comment (twiecki) [reply hddm/models/hddm_regression.py:25]: Did you overload `__init__()`?

Comment (madslupe) [reply hddm/models/hddm_regression.py]: I've moved the wfpt_parents['alpha'] to inside the HDDMrlRegressor class and removed alpha from all other likelihoods. And the HDDMrlRegressor works as before, i.e. allows regressing against alpha.

File: hddm/models/hddm_info.py [hunk 4]
```diff
@@ -126,6 +126,8 @@ def _create_stochastic_knodes_info(self, include):
             knodes.update(self._create_family_normal_normal_hnormal('v', value=2, g_mu=2, g_tau=3**-2, std_std=2))
         if 't' in include:
             knodes.update(self._create_family_gamma_gamma_hnormal('t', g_mean=.4, g_std=0.2, value=0.001, std_std=1, std_value=0.2))
+        if 'alpha' in include:
```
Comment (twiecki) [inline hddm/models/hddm_info.py:129]: This should also be pushed into your own class.

File: hddm/models/hddm_regression.py [hunk 5]
```diff
@@ -103,9 +103,7 @@ class HDDMRegressor(HDDM):
 
     def __init__(self, data, models, group_only_regressors=True, keep_regressor_trace=False, **kwargs):
         """Instantiate a regression model.
-
```
Comment (twiecki) [inline hddm/models/hddm_regression.py:106]: Why did you remove these empty lines?

Comment (madslupe) [reply hddm/models/hddm_regression.py:106]: Sorry, that was not on purpose. Will fix it.

File: hddm/models/hddm_info.py (hunk 4 above)
Comment (madslupe) [reply hddm/models/hddm_info.py:129]: I haven't found a good way to this, while still being able to regress against the new parameter. The problem is that if I update the knodes with alpha after the following line it doesn't create alpha_Intercept:

        # Create all stochastic knodes except for the ones that we want to replace
        # with regressors.
        knodes = super(HDDMrlRegressor, self)._create_stochastic_knodes(include.difference(self.reg_outcomes))

Any ideas how to add alpha inside my own class but before creating knodes?

File: hddm/models/base.py [hunk 6]
```diff
@@ -703,8 +703,8 @@ def __init__(self, data, bias=False, include=(),
         if bias:
             self.include.add('z')
 
-        possible_parameters = ('v', 'a', 't', 'z', 'st', 'sz', 'sv', 'p_outlier','dual_alpha')
```
Comment (twiecki) [inline hddm/models/base.py:706]: I think we need `dual_alpha` elsewhere, can you just add `alpha` here?

File: hddm/models/hddm_info.py (hunk 4 above)
Comment (twiecki) [reply hddm/models/hddm_info.py:129]: in your overloaded method you should be able to just do:
```
knodes = super(HDDMrlRegressor, self)._create_stochastic_knodes(includes)
if 'alpha' in include:            
    knodes.update(self._create_family_normal_normal_hnormal('alpha', value=0, g_tau=50**-2, std_std=10))
return knodes
```

File: hddm/models/base.py (hunk 6 above)
Comment (madslupe) [reply hddm/models/base.py:706]: It's actually not used anymore, instead a parameter called pos_alpha is directly created in the HDDMrl and Hrl classes if dual=True.

File: hddm/models/hddm_info.py (hunk 4 above)
Comment (madslupe) [reply hddm/models/hddm_info.py:129]: Sorry, there's probably something trivial about the overloaded method I'm not getting. I've tried adding the knode for alpha at line 242 in HDDMrlRegressor.py, but then it does not create alpha_Intercept if (alpha~BOLD). Where do you suggest I add this?

knodes = super(HDDMrlRegressor, self)._create_stochastic_knodes(includes)
if 'alpha' in include:            
    knodes.update(self._create_family_normal_normal_hnormal('alpha', value=0, g_tau=50**-2, std_std=10))
return knodes

File: hddm/models/hddm_rl_regression.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+            model['link_func'] = lambda x: x
+        super(HDDMrlRegressor, self).__setstate__(d)
+
+    def _create_wfpt_knode(self, knodes):
+        wfpt_parents = self._create_wfpt_parents_dict(knodes)
+        wfpt_parents['alpha'] = knodes['alpha_bottom']
+        return Knode(self.wfpt_reg_class, 'wfpt', observed=True,
+                     col_name=['split_by', 'feedback', 'response', 'rt', 'q_init'],
+                     reg_outcomes=self.reg_outcomes, **wfpt_parents)
+
```
Comment (twiecki) [inline hddm/models/hddm_rl_regression.py:237]: Here you add a new method called `_create_stochastic_knodes` with the code I posted.

File: hddm/models/hddm_rl_regression.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+            # Apply design matrix to input data
+            if design_matrix.shape[1] != params.shape[1]:
+                raise NotImplementedError('Missing columns in design matrix. You need data for all conditions for all subjects.')
+            predictor = link_func(pd.DataFrame((design_matrix * params).sum(axis=1), index=data.index))
+
+            return pd.DataFrame(predictor, index=data.index)
+
+        return self.pymc_node(func, kwargs['doc'], name, parents=parents, trace=self.keep_regressor_trace)
+
+class HDDMrlRegressor(HDDM):
```
Comment (twiecki) [inline hddm/models/hddm_rl_regression.py:102]: You might want to inherit from HDDMInfo instead.

File: hddm/models/hddm_rl_regression.py (hunk 7 above)
Comment (madslupe) [reply hddm/models/hddm_rl_regression.py:237]: Adding the following here while also removing the creation of alpha in base.py results in an error. Note I also keep the original create_stochastic_knodes (starting at line 238 here).

 def _create_stochastic_knodes(self, include):
        knodes = super(HDDMrlRegressor, self)._create_stochastic_knodes(includes)
        if 'alpha' in include:            
            knodes.update(self._create_family_normal_normal_hnormal('alpha', value=0, g_tau=50**-2, std_std=10))
        return knodes


---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-5-5bd3dbb5434f> in <module>
----> 1 m = hddm.HDDMrlRegressor(data,'a ~ neural_regressor',include='alpha')
      2 m.sample(50,burn=30,dbname='traces.db',db='pickle')
      3 m.print_stats()

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/hddm_rl_regression.py in __init__(self, data, models, group_only_regressors, keep_regressor_trace, **kwargs)
    206         self.wfpt_reg_class = deepcopy(wfpt_reg_like)
    207 
--> 208         super(HDDMrlRegressor, self).__init__(data, **kwargs)
    209 
    210         # Sanity checks

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/hddm_info.py in __init__(self, *args, **kwargs)
    111         self.is_informative = kwargs.pop('informative', True)
    112 
--> 113         super(HDDM, self).__init__(*args, **kwargs)
    114 
    115     def _create_stochastic_knodes(self, include):

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/base.py in __init__(self, data, bias, include, wiener_params, p_outlier, **kwargs)
    727         self.wfpt_class = hddm.likelihoods.generate_wfpt_stochastic_class(wp, cdf_range=self.cdf_range)
    728 
--> 729         super(HDDMBase, self).__init__(data, **kwargs)
    730 
    731     def __getstate__(self):

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/base.py in __init__(self, data, **kwargs)
     38         self.std_depends = kwargs.pop('std_depends', False)
     39 
---> 40         super(AccumulatorModel, self).__init__(data, **kwargs)
     41 
     42 

~/anaconda/envs/py36/lib/python3.6/site-packages/kabuki/hierarchical.py in __init__(self, data, is_group_model, depends_on, trace_subjs, plot_subjs, plot_var, group_only_nodes)
    346         self.db = None
    347 
--> 348         self._setup_model()
    349 
    350     def _setup_model(self):

~/anaconda/envs/py36/lib/python3.6/site-packages/kabuki/hierarchical.py in _setup_model(self)
    350     def _setup_model(self):
    351         # create knodes (does not build according pymc nodes)
--> 352         self.knodes = self.create_knodes()
    353 
    354         #add data to knodes

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/base.py in create_knodes(self)
    759     def create_knodes(self):
    760         knodes = self._create_stochastic_knodes(self.include)
--> 761         knodes['wfpt'] = self._create_wfpt_knode(knodes)
    762 
    763         return list(knodes.values())

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/hddm_rl_regression.py in _create_wfpt_knode(self, knodes)
    231     def _create_wfpt_knode(self, knodes):
    232         wfpt_parents = self._create_wfpt_parents_dict(knodes)
--> 233         wfpt_parents['alpha'] = knodes['alpha_bottom']
    234         return Knode(self.wfpt_reg_class, 'wfpt', observed=True,
    235                      col_name=['split_by', 'feedback', 'response', 'rt', 'q_init'],

KeyError: 'alpha_bottom'

Comment (twiecki) [reply hddm/models/hddm_rl_regression.py:237]: is `alpha` in includes?

Comment (madslupe) [reply hddm/models/hddm_rl_regression.py:237]: Yes

Comment (twiecki) [reply hddm/models/hddm_rl_regression.py:237]: Then I'm not sure, maybe try with a debugger to see what's happening.

Comment (madslupe) [reply hddm/models/hddm_rl_regression.py:237]: I'll try to figure out how to solve it. Related to that, why shouldn't the knode for alpha be created in  hddm_info.py? It doesn't seem to influence the other classes.

Comment (twiecki) [reply hddm/models/hddm_rl_regression.py:237]: Well we could and it's probably fine for this. These inheritance abstractions exist exactly for this purpose, though.

File: hddm/models/hddm_rl_regression.py [hunk 9]
```diff
@@ -228,8 +228,14 @@ def __setstate__(self, d):
             model['link_func'] = lambda x: x
         super(HDDMrlRegressor, self).__setstate__(d)
 
+    def _create_stochastic_knodes_rl(self, include):
```
Comment (twiecki) [inline hddm/models/hddm_rl_regression.py:231]: 👍

File: hddm/models/hddm_info.py [hunk 10]
```diff
@@ -102,10 +102,10 @@ class HDDM(HDDMBase):
     def __init__(self, *args, **kwargs):
         self.slice_widths = {'a':1, 't':0.01, 'a_std': 1, 't_std': 0.15, 'sz': 1.1, 'v': 1.5,
                              'st': 0.1, 'sv': 3, 'z_trans': 0.2, 'z': 0.1,
-                             'p_outlier':1., 'v_std': 1,'alpha':3.5,'dual_alpha':3.5}
+                             'p_outlier':1., 'v_std': 1, 'alpha':3.5, 'dual_alpha':3.5}
         self.emcee_dispersions = {'a':1, 't': 0.1, 'a_std': 1, 't_std': 0.15, 'sz': 1.1, 'v': 1.5,
                                   'st': 0.1, 'sv': 3, 'z_trans': 0.2, 'z': 0.1,
-                                  'p_outlier':1., 'v_std': 1,'alpha':3.5,'dual_alpha':3.5}
+                                  'p_outlier':1., 'v_std': 1, 'alpha':3.5, 'dual_alpha':3.5}
```
Comment (twiecki) [inline hddm/models/hddm_info.py:108]: can remove `dual_alpha` here and above.

File: hddm/models/hddm_info.py [hunk 11]
```diff
@@ -102,10 +102,10 @@ class HDDM(HDDMBase):
     def __init__(self, *args, **kwargs):
         self.slice_widths = {'a':1, 't':0.01, 'a_std': 1, 't_std': 0.15, 'sz': 1.1, 'v': 1.5,
                              'st': 0.1, 'sv': 3, 'z_trans': 0.2, 'z': 0.1,
-                             'p_outlier':1., 'v_std': 1,'alpha':3.5,'dual_alpha':3.5}
+                             'p_outlier':1., 'v_std': 1, 'alpha':3.5, 'dual_alpha':3.5}
```
Comment (twiecki) [inline hddm/models/hddm_info.py:105]: Are you sure that this is the right slice_widths? This depends on the expected width of the posterior, and 3.5 is very wide.

File: hddm/models/hddm_info.py (hunk 10 above)
Comment (madslupe) [reply hddm/models/hddm_info.py:108]: Yes, or actually, I'll replace it with pos_alpha, which is the parameter being created if dual is set to True for HDDMrl and Hrl.

File: hddm/models/hddm_info.py (hunk 11 above)
Comment (madslupe) [reply hddm/models/hddm_info.py:105]: Thanks for catching this. I've now set them to be 1.5,  same value as v.

**Comment by madslupe** [comment]
Thomas, you think this is ready for merging?

File: hddm/models/hddm_regression.py (hunk 5 above)
Comment (twiecki) [reply hddm/models/hddm_regression.py:106]: Still needs fixing.

File: hddm/models/hddm_regression.py [hunk 12]
```diff
@@ -152,9 +152,10 @@ def __init__(self, data, models, group_only_regressors=True, keep_regressor_trac
                 ```python
                 HDDMRegressor(data, 'v ~ C(condition)')
                 ```
+
             This will lead to estimation of 'v_Intercept' for cond1
             and v_C(condition)[T.cond2] for cond1+cond2.
-
+
```
Comment (twiecki) [inline hddm/models/hddm_regression.py:158]:
```suggestion

```

**Comment by twiecki** [comment]
Nice one!

**Comment by madslupe** [comment]
Thanks! Should I update to 0.7.6 as well?

**Comment by twiecki** [comment]
pypi? yes!
